odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1516 @@
+
+        Notes
+        -----
+        The convex conjugate functional of a convex functional :math:`f(x)`,
+        defined on a Hilber space, is defined as the functional
+
+        .. math::
+
+            f^*(x^*) = \\sup_{x} \{ \\langle x^*,x \\rangle - f(x)  \}.
+
+        See, e.g., [Lue1969]_, [Roc1970]_.
+        """
+        raise NotImplementedError
+
+    def derivative(self, point):
+        """Returns the derivative operator in the given point.
+
+        This function returns the linear operator
+
+            ``x --> <x, grad_f(point)>``,
+
+        where ``grad_f(point)`` is the gradient of the functional in the point
+        ``point``.
+
+        Parameters
+        ----------
+        point : `LinearSpaceVector`
+            The point in which the gradient is evaluated.
+
+        Returns
+        -------
+        out : `DerivativeOperator`
+            The linear operator that maps ``x --> <x, grad_f(point)>``.
+        """
+        functional = self
+
+        # TODO: Update doc for this operator/functional
+        # Also: it is a functional, not an operator, right?
+        class DerivativeOperator(Functional):
+            def __init__(self):
+                super().__init__(functional.domain, linear=True)
+
+            self.point = point
+
+            def _call(self, x):
+                return x.inner(functional.gradient(point))
+
+        return DerivativeOperator()
+
+    def translate(self, translation):
+        """Creates a translation of the functional.
+
+        For a given functional ``f`` and an element ``translation`` in the
+        domain of ``f``, this operation creates the functional
+        ``f(. - translation)``.
+
+        Parameters
+        ----------
+        translation : `LinearSpaceVector`
+            Element in the domain of the functional
+
+        Returns
+        -------
+        out : `TranslatedFunctional`
+            The functional ``f(. - translation)``
+        """
+        if translation not in self.domain:
+            raise TypeError('the translation {} is not in the domain of the'
+                            'functional {!r}'.format(translation, self))
+
+        return TranslatedFunctional(self, translation)
+
+    def __mul__(self, other):
+        """Return ``self * other``.
+
+        If ``other`` is an operator, this corresponds to composition with the
+        operator:
+
+            ``func * op <==> (x --> func(op(x))``
+
+        If ``other`` is a scalar, this corresponds to right multiplication of
+        scalars with functionals:
+
+            ``func * scalar <==> (x --> func(scalar * x))``
+
+        If ``other`` is a vector, this corresponds to right multiplication of
+        vectors with functionals:
+
+            ``func * vector <==> (x --> func(vector * x))``
+
+        Note that left and right multiplications are generally different.
+
+        Parameters
+        ----------
+        other : {`Operator`, `LinearSpaceVector`, scalar}
```

Comment:
The `{a, b, c}` style is only intended for a fixed range of *values*, not *types*. If we have used it in other places, it's wrong, too. Use "or" to separate them, and if it gets too long, leave out the type completely and list the types with corresponding behavior, just as you did here.